usrp_drivers/n200/n200_driver.cpp

```diff
@@ -24,271 +24,339 @@
-      std::cout << "TRANSMIT Samples to send " << num_samps_to_send << std::endl;
+      DEBUG_MSG("TRANSMIT Samples to send " << num_samps_to_send);
 
-      num_samps_sent = tx_stream->send(
-        samples, num_samps_to_send, md.get_md());
-
-      std::cout << "TRANSMIT Samples sent " << num_samps_sent << std::endl;
+      //Send behaviour can be found in UHD docs
+      num_samps_sent = tx_stream->send( // REVIEW #37 How do we know if it's timed out? Does it do the fragmentation of packets internal to the send(...) if above the max num samps value?
+        samples, num_samps_to_send, md.get_md()); // REVIEW #0 Should we set the timeout to a reasonable value given our timing constraints? 100ms would be way too long to recover this pulse sequence. Maybe we can make a calculation based on how long this pulse is and set the timeout slightly beyond that - or beyond atten_time_low (at which point you should be in this loop for the next pulse) We can also get_max_num_samps to find out if this call will fragment into multiple packets, not sure of the behaviour there, does it return before the timeout??
+                                                  // REPLY lets talk about this.
+      DEBUG_MSG("TRANSMIT Samples sent " << num_samps_sent);
 
       md.set_start_of_burst(false);
       md.set_has_time_spec(false);
     }
 
     md.set_end_of_burst(true);
-    tx_stream->send("", 0, md.get_md());
+    tx_stream->send("", 0, md.get_md()); // REVIEW #43 If we know that we are under the max num samps value, then we could send start of burst and end of burst in the same send(...) call
+                                         // REPLY would have to test that, but im not sure what that would gain. This is how its done in UHD examples.
+    auto end_send = std::chrono::steady_clock::now();
 
-    end = std::chrono::steady_clock::now();
+    time_diff = std::chrono::duration_cast<std::chrono::microseconds>
+                                                          (end_send - begin_send).count();
+    DEBUG_MSG("TRANSMIT time to send to USRP: " << time_diff << "us");
 
-    std::cout << "TRANSMIT time to send to USRP: "
-      << std::chrono::duration_cast<std::chrono::microseconds>
-                                                          (end - begin).count()
-      << "us" << std::endl;
+    auto timing_start = std::chrono::steady_clock::now();
 
-    std::chrono::steady_clock::time_point timing_start = std::chrono::steady_clock::now();
+    //Set high speed IO timing on the USRP now.
+    usrp_d.clear_command_times(); // REVIEW #0 The GPIO timing setup should be moved to above the samples sending, in case we timeout on sending samples (i.e. we sent half of them) then the RF waveform would be transmitting without timing signals, bad idea. Alternatively, what if there was a 4th thread for timing specifically?
+                                             // REPLY I tried this. It doesnt work. If this is a problem then we need a circuit for this.
```

Comment:
Therefore we need to know that the GPIO timing is actually happening somehow, otherwise we'll have no power going out, or if the timing was late and TR went high in the middle of a pulse we would have no ramp-up (or ramp-down in the other case)

Reply:
yes agreed.